### PR TITLE
Add firestore adapter - Clean Branch

### PR DESCRIPTION
**Pull request by vihangpatel-hr**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@hackerrank/firepad",
   "description": "Collaborative text editing powered by Firebase",
-  "version": "0.4.1-beta",
+  "version": "0.5.1-alpha",
```
Comment (karthikdivi) [inline package.json:4]: can we use 0.7.0-beta? because the 0.5.x, and 0.6.x are already used from the rolled-out branches and from our previous PR

**Review by karthikdivi — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (vihangpatel-hr) [reply package.json]: Fixed.

**Review by karthikdivi — APPROVED** [review]

File: src/firestore-adapter.ts [hunk 2]
```diff
@@ -0,0 +1,859 @@
...
+   * @param revisionData - Text Operation and metadata in JSON format.
+   * @param callback - Success/Error callback handler.
+   */
+  protected _doTransaction(
+    revisionId: string,
+    revisionData: FirebaseOperationDataType,
+    callback: SendOperationCallbackType
+  ): void {
+    this._firestoreRef!.firestore.runTransaction(async (t) => {
+      t.set(
```
Comment (aditya211935) [inline src/firestore-adapter.ts:568]: In the transaction, we need to check if data already exists in `revisionId`. If it already exists, we need to cancel the transaction.
Two clients can simultaneously push operation with same revisionId (let's say A1). Then if both are allowed to succeed, one client's operation will be overridden by another client. Ref: https://github.com/interviewstreet/firepad-x/blob/236ece5f29b1790e03bd6e3c28cfcc34c91f56ca/src/firebase-adapter.ts#L474-L476

**Review by aditya211935 — CHANGES_REQUESTED** [review]

File: src/firestore-adapter.ts [hunk 3]
```diff
@@ -0,0 +1,859 @@
...
+          pending[revisionId]
+        );
+      } else {
+        this._document = this._document!.compose(revision.operation);
+
+        if (this._sent && revisionId === this._sent.id) {
+          // We have an outstanding change at this revision id.
+          if (
+            this._sent.op.equals(revision.operation) &&
+            revision.author == this._userId
```
Comment (aditya211935) [inline src/firestore-adapter.ts:431]: Can we add the check from this PR (https://github.com/interviewstreet/firepad-x/pull/69) here?

**Review by aditya211935 — CHANGES_REQUESTED** [review]

File: src/firestore-adapter.ts (hunk 2 above)
Comment (MrSurana) [reply src/firestore-adapter.ts]: Added the check.

File: src/firestore-adapter.ts (hunk 3 above)
Comment (MrSurana) [reply src/firestore-adapter.ts:431]: Lets first go ahead with firestore-only changes, and have a separate release for this change.

**Review by aditya211935 — APPROVED** [review]

**Review by MrSurana — APPROVED** [review]

**Pull request merged**